docs/src/adr/015adr-trace.md

```diff
@@ -0,0 +1,403 @@
+      - requires the knowledge of Apalache IR.
+
+## Solution
+
+In this ADR, we propose a very simple format that represents executions of
+state machines that follows the concepts of TLA+ and yet avoids complexity of
+TLA+. It is so simple that we call it *"Informal Trace Format"* (ITF).
+(Obviously, it is formal enough to be machine-readable).  By convention, the
+files in this format should end with the extension `.itf.json`.
+
+### The ITF Format
+
+**Trace object**. A trace in ITF is a JSON object:
+
+```js
+{
+  "#meta": <optional object>,
+  "params": <optional array of strings>,
+  "vars":  <array of strings>,
+  "states": <array of states>,
+  "loop": <optional int>
+}
+```
+
+The field `#meta` is an arbitrary JSON object, whose purpose is to provide
+the reader with additional comments about the trace. For example, it may look
+like:
+
+```js
+  "#meta": {
+    "description": "Generated by Apalache",
+    "source": "MissionariesAndCannibalsTyped.tla"
+  }
+```
+
+The optional field `params` is an array of names that must be set in the
+initial state (if there are any parameters). The parameters play the same role
+as `CONSTANTS` in TLA+. For example, the field may look like:
+
+```js
+  "params": [ "Missionaries", "Cannibals" ]
+```
+
+The field `vars` is an array of names that must be set in every state.
+For example, the field may look like:
+
+```js
+  "vars": [ "bank_of_boat", "who_is_on_bank" ]
+```
+
+The field `states` is an array of state objects (see below). For example,
+the field may look like:
+
+```js
+  "states": [ <state0>, <state1>, <state2> ]
+```
+
+The optional field `loop` specifies the index of the state (in the array of
+states) that starts the loop. The loop ends in the last state. For example,
+the field may look like:
+
+```js
+  "loop": 1
+```
+
+**State object**. A state is a JSON object:
+
+```js
+  {
+    "#meta": <optional object>,
+    "<var1>": <expr>,
+    ...
+    "<varN>": <expr>
+  }
+```
+
+As in the trace object, the field `#meta` may be an arbitrary object.
+Different tools may use this object to write their metadata into this object.
+
+The names `<var1>, ..., <varN>` are the names of the variables that are
+specified in the field `vars`. Each state must define a value for every specified variable. The syntax of `<expr>` is specified below. 
+
+**Expressions.** As usual, expressions are inductively defined. An expression
+`<expr>` is one of the following:
+
+ 1. A JSON string literal, e.g., `"hello"`. TLA+ strings are written as strings
+    in this format.
+
+ 1. A JSON integer, e.g., 123. According to [RFC7159][], JSON integers
+    must be in the range: `[-(2**53)+1, (2**53)-1]`. Integers in this
+    range *may be* written as JSON integers.
+
+ 1. A big integer of the following form: `{ "#bigint": "[-][0-9]+" }`.
+    We are using this format, as many JSON parsers impose limits on
+    integer values, see [RFC7159][]. Big and small integers *may be*
+    written in this format.
```

Comment:
I'm slightly unsure about this complication, since it introduces two different ways of representing what should be, iiuc, understood as just a single kind of value. If any program that consumes this data is going to have to support bigints  anyhow, should we perhaps only have `{"#int":  "[-][0-9]+" }` as the supported value for integer-expressions?

Reply:
True. However, if we put focus on usability in this format, not on the easy of writing a parser, it makes sense to write small integers without typing too much.

Reply:
Imo, this is a shortsighted view of usability. It’s optimizing for the simple case instead of regularity and generality. It means users have to study documentation and memorize which values get encoded which way and why.

in general, and I know I keep saying this, optimizing data exchange formats for human readability is a bad idea. If you want a readable view, just implement a small script to print a markdown table or a dot graph or an avg or whatever. 

Every time I’ve had to work with a data format where authors compromised the regularity of the machine encoding so that is has nice features as ascii art I’ve found it to slow me down and make things more tediuous (YAML in general is a great example).

Reply:
I see your point. However, we already have a perfect machine-readable format. I really hope that JSON fixes this issue of bounded integers in the future. For instance, TypeScript supports big integers, but they are not serialized into JSON very well.

Reply:
I have added a discussion on that, so we can come back to your point later :-)